### PR TITLE
Fix async writing of files to jobstore in caching (resolves #966)

### LINKED ISSUE #966: Wrong permissions on file in fileJobStore
```
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Obtained lock on file /tmp/toil-bf91efaf-3d0a-441d-8e55-40a040a201d7/cache-bf91efaf-3d0a-441d-8e55-40a040a201d7/.cacheLock
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Released lock
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Obtained lock on file /tmp/toil-bf91efaf-3d0a-441d-8e55-40a040a201d7/cache-bf91efaf-3d0a-441d-8e55-40a040a201d7/.cacheLock
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Released lock
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Obtained lock on file /tmp/toil-bf91efaf-3d0a-441d-8e55-40a040a201d7/cache-bf91efaf-3d0a-441d-8e55-40a040a201d7/.cacheLock
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Released lock
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Obtained lock on file /tmp/toil-bf91efaf-3d0a-441d-8e55-40a040a201d7/cache-bf91efaf-3d0a-441d-8e55-40a040a201d7/.cacheLock
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Cache miss on file with ID 'R/q/tmp4jyd3q.tmp'.
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Obtained lock on file /tmp/toil-bf91efaf-3d0a-441d-8e55-40a040a201d7/cache-bf91efaf-3d0a-441d-8e55-40a040a201d7/.cacheLock
T/O/jobugCtyo:  INFO:toil.job:CACHE: Read file with ID 'R/q/tmp4jyd3q.tmp' from the cache.
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Released lock
T/O/jobugCtyo:  DEBUG:toil.job:CACHE: Released lock
T/O/jobugCtyo:  DEBUG:root:Running the command: cactus_workflow_getFlowers CRITICAL '<st_kv_database_conf type="kyoto_tycoon">
T/O/jobugCtyo:                          <kyoto_tycoon database_dir="/tmp/toil-bf91efaf-3d0a-441d-8e55-40a040a201d7/tmp4egzRz/2087b518-a4bc-4613-9299-5c6b2ac71f11/tL3udCx/cactusDB" database_name="Anc2.kch" host="thinkpad" in_me
mory="1" port="1979" snapshot="0" />
T/O/jobugCtyo:                  </st_kv_database_conf>' 0 100000000 50000000
T/O/jobugCtyo:  Exception in thread Thread-25:
T/O/jobugCtyo:  Traceback (most recent call last):
T/O/jobugCtyo:    File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
T/O/jobugCtyo:      self.run()
T/O/jobugCtyo:    File "/usr/lib/python2.7/threading.py", line 754, in run
T/O/jobugCtyo:      self.__target(*self.__args, **self.__kwargs)
T/O/jobugCtyo:    File "/home/alden/progressiveCactus-new/python/local/lib/python2.7/site-packages/toil/job.py", line 571, in asyncWrite
T/O/jobugCtyo:      with jobStore.updateFileStream(jobStoreFileID) as outputFileHandle:
T/O/jobugCtyo:    File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
T/O/jobugCtyo:      return self.gen.next()
T/O/jobugCtyo:    File "/home/alden/progressiveCactus-new/python/local/lib/python2.7/site-packages/toil/jobStores/fileJobStore.py", line 255, in updateFileStream
T/O/jobugCtyo:      with open(self._getAbsPath(jobStoreFileID), 'w') as f:
T/O/jobugCtyo:  IOError: [Errno 13] Permission denied: '/home/alden/alignments/toil/work/toil/tmp/R/q/tmp4jyd3q.tmp'
```

I tried to make a test case for this outside of cactus but wasn't able to recreate it exactly. It looks like the file is ending up with 444 permissions - I'm not sure if this is ever intended to occur or not. This appears to happen with graphs like this:

A --> C 
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; |
B &nbsp;&nbsp;&nbsp;&nbsp; D

where A writes a file to the file store, B downloads it with mutable=True, then modifies it and writes it back to the file store with cleanup=False, and D downloads it. The test case I made with the above graph doesn't trigger the asyncWrite error, but the file store ID does have 444 permissions in job D.

### PR DESCRIPTION
**Pull request by arkal**
resolves #966 

asyncWrite is now a class method and CachedFileStore overloads it with a cache-friendly version

### CODE REVIEW

**Comment by arkal** [comment]
Jenkins, test this please.

**Comment by arkal** [comment]
Jenkins, test this please.

**Comment by arkal** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
Pending verification by author of #966.

**Comment by hannes-ucsc** [comment]
I would like this rebased, too much has happened on master in the past two days.

File: src/toil/job.py [hunk 1]
```diff
@@ -998,6 +1000,52 @@ def __init__(self, jobStore, jobWrapper, localTempDir, inputBlockFn):
...
+                        harbingerFile.write(str(os.getpid()))
+                    os.rename(harbingerFileName + '.tmp', harbingerFileName)
+                    # We pass in a fileHandle, rather than the file-name, in case
+                    # the file itself is deleted. The fileHandle itself should persist
+                    # while we maintain the open file handle
+                    with self.jobStore.updateFileStream(jobStoreFileID) as outputFileHandle:
+                        bufferSize=1000000  # TODO: This buffer number probably needs to be
+                        # modified/tuned
+                        while 1:
+                            copyBuffer = inputFileHandle.read(bufferSize)
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Use shutil.copyfileobj()

**Comment by arkal** [comment]
OK check out [build 1492](http://jenkins.cgcloud.info/job/toil-pull-requests/1492/console).
FileJobStore fails with the expected error.
AwsJobStore is slightly different. It is failing at line 1005 in aws/jobStore.py mostly because self.content is None. This is because job C in jobCacheTest has asked for the file before the async write process in B could finish writing to the job store (possibly hasn't even started). This will all be addressed in the next commit

**Comment by arkal** [comment]
[Build 1493](http://jenkins.cgcloud.info/job/toil-pull-requests/1493/) shows that the tests pass in both AWS and File job stores.

@adderan this should fix your issue. It looks a little complicated setting up cactus so it would be great if you could test this on your cactus script.

**Comment by arkal** [comment]
@hannes-ucsc Alden has  confirmed this works on Cactus! This is ready for review

**Comment by hannes-ucsc** [comment]
Build failure tracked in #760.

File: src/toil/job.py [hunk 2]
```diff
@@ -998,6 +1000,51 @@ def __init__(self, jobStore, jobWrapper, localTempDir, inputBlockFn):
 
             self._setupCache()
 
+        # Cache-friendly version if asyncWrite
+        def asyncWrite(self):
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Mention in a comment that this duplicates code and which code.

File: src/toil/job.py [hunk 3]
```diff
@@ -998,6 +1000,51 @@ def __init__(self, jobStore, jobWrapper, localTempDir, inputBlockFn):
...
+                    # is off-limits till the async update is complete.
+                    cachedFileName = self.encodedFileID(jobStoreFileID)
+                    harbingerFileName = self.getHarbingerFileName(cachedFileName=cachedFileName)
+                    with open(harbingerFileName + '.tmp', 'w') as harbingerFile:
+                        harbingerFile.write(str(os.getpid()))
+                    os.rename(harbingerFileName + '.tmp', harbingerFileName)
+                    # We pass in a fileHandle, rather than the file-name, in case
+                    # the file itself is deleted. The fileHandle itself should persist
+                    # while we maintain the open file handle
+                    with self.jobStore.updateFileStream(jobStoreFileID) as outputFileHandle:
```
Comment (hannes-ucsc) [inline src/toil/job.py]: shutil.copyfileobj

File: src/toil/job.py [hunk 4]
```diff
@@ -548,47 +548,49 @@ def __init__(self, jobStore, jobWrapper, localTempDir, inputBlockFn):
...
+                            raise RuntimeError("The termination flag is set, exiting")
+                        continue
+                    # Normal termination condition is getting None from queue
+                    if args is None:
+                        break
+                    inputFileHandle, jobStoreFileID = args
+                    # We pass in a fileHandle, rather than the file-name, in case
+                    # the file itself is deleted. The fileHandle itself should persist
+                    # while we maintain the open file handle
+                    with self.jobStore.updateFileStream(jobStoreFileID) as outputFileHandle:
```
Comment (hannes-ucsc) [inline src/toil/job.py]: shutil.copyfileobj

Comment (hannes-ucsc) [inline src/toil/job.py]: resolved

**Comment by hannes-ucsc** [comment]
Ran out of domains/buckets again.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

File: src/toil/job.py [hunk 5]
```diff
@@ -1930,6 +1977,44 @@ def updateJobReqs(self, fileSize, actions):
...
+                """
+                Returns the harbinger file name for a cahed file, or for a job store ID
+                :param class outer: The 'self' object of the outer class
+                :param str fileStoreID:
+                :param str cachedFileName:
+                :return: Harbinger file name
+                :rtype: str
+                """
+                assert (fileStoreID and not cachedFileName) or (cachedFileName and not fileStoreID)
+                if fileStoreID:
```
Comment (hannes-ucsc) [inline src/toil/job.py]: is not None

File: src/toil/job.py [hunk 6]
```diff
@@ -1930,6 +1977,44 @@ def updateJobReqs(self, fileSize, actions):
...
+            def __init__(self, outer, fileStoreID=None, cachedFileName=None):
+                """
+                Returns the harbinger file name for a cahed file, or for a job store ID
+                :param class outer: The 'self' object of the outer class
+                :param str fileStoreID:
+                :param str cachedFileName:
+                :return: Harbinger file name
+                :rtype: str
+                """
+                assert (fileStoreID and not cachedFileName) or (cachedFileName and not fileStoreID)
```
Comment (hannes-ucsc) [inline src/toil/job.py]: `assert (fileStoreID is None) != (cachedFileName is None)` is shorter and more specific.

I generally don't like equating None and False-ish values. It's a source for bugs.

File: src/toil/job.py [hunk 7]
```diff
@@ -1930,6 +1977,44 @@ def updateJobReqs(self, fileSize, actions):
...
 
+        class HarbingerFile(object):
+            """
+            A class to handle the various harbinger file operations.
+            """
+            def __init__(self, outer, fileStoreID=None, cachedFileName=None):
+                """
+                Returns the harbinger file name for a cahed file, or for a job store ID
+                :param class outer: The 'self' object of the outer class
+                :param str fileStoreID:
```
Comment (hannes-ucsc) [inline src/toil/job.py]: document

File: src/toil/job.py [hunk 8]
```diff
@@ -1930,6 +1977,44 @@ def updateJobReqs(self, fileSize, actions):
...
+                :param str cachedFileName:
```
Comment (hannes-ucsc) [inline src/toil/job.py]: document

File: src/toil/job.py [hunk 9]
```diff
@@ -1930,6 +1977,44 @@ def updateJobReqs(self, fileSize, actions):
...
+                """
+                assert (fileStoreID and not cachedFileName) or (cachedFileName and not fileStoreID)
+                if fileStoreID:
+                    cachedFileName = outer.encodedFileID(fileStoreID)
+                self.harbingerFileName = '/.'.join(os.path.split(cachedFileName)) + '.harbinger'
+
+            def write(self):
+                """
+                Atomically write the pid of the current process to the harbinger file.
+                :return:
```
Comment (hannes-ucsc) [inline src/toil/job.py]: redundant

File: src/toil/job.py [hunk 10]
```diff
@@ -1930,6 +1977,44 @@ def updateJobReqs(self, fileSize, actions):
...
+                """
+                with open(self.harbingerFileName + '.tmp', 'w') as harbingerFile:
+                    harbingerFile.write(str(os.getpid()))
+                os.rename(self.harbingerFileName + '.tmp', self.harbingerFileName)
+                # Make this File read only to prevent overwrites
+                os.chmod(self.harbingerFileName, 0444)
```
Comment (hannes-ucsc) [inline src/toil/job.py]: first chmod, then rename

File: src/toil/job.py [hunk 11]
```diff
@@ -1930,6 +1977,44 @@ def updateJobReqs(self, fileSize, actions):
...
+                :param class outer: The 'self' object of the outer class
+                :param str fileStoreID:
+                :param str cachedFileName:
+                :return: Harbinger file name
+                :rtype: str
+                """
+                assert (fileStoreID and not cachedFileName) or (cachedFileName and not fileStoreID)
+                if fileStoreID:
+                    cachedFileName = outer.encodedFileID(fileStoreID)
+                self.harbingerFileName = '/.'.join(os.path.split(cachedFileName)) + '.harbinger'
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Wow. I had to read this a few times. But ok.

File: src/toil/job.py (hunk 6 above)
Comment (hannes-ucsc) [inline src/toil/job.py]: and document the constraint

File: src/toil/job.py [hunk 12]
```diff
@@ -1930,6 +1977,44 @@ def updateJobReqs(self, fileSize, actions):
             def isPopulated(self):
                 return self.__dict__ != {}
 
+        class HarbingerFile(object):
+            """
+            A class to handle the various harbinger file operations.
+            """
+            def __init__(self, outer, fileStoreID=None, cachedFileName=None):
```
Comment (hannes-ucsc) [inline src/toil/job.py]: rename outer to fileStore

File: src/toil/job.py [hunk 13]
```diff
@@ -1930,6 +1977,44 @@ def updateJobReqs(self, fileSize, actions):
             def isPopulated(self):
                 return self.__dict__ != {}
 
+        class HarbingerFile(object):
+            """
+            A class to handle the various harbinger file operations.
```
Comment (hannes-ucsc) [inline src/toil/job.py]:
```
Represents the placeholder file that harbinges the arrival of a local copy of a file in the job store.
```

Is that better?

File: src/toil/job.py [hunk 14]
```diff
@@ -1176,13 +1225,13 @@ def readGlobalFile(self, fileStoreID, userPath=None, cache=True, mutable=None):
                 # FileStoreID exists.  If it does, the wait and periodically check for the removal
                 # of the file and the addition of the completed download into cache of the file by
                 # the other job. Then we link to it.
-                elif fileIsLocal and os.path.exists(harbingerFileName):
+                elif fileIsLocal and harbingerFile.exists():
                     logger.info('CACHE: Waiting for another worker to download file with ID %s.'
                                 % fileStoreID)
-                    while os.path.exists(harbingerFileName):
+                    while harbingerFile.exists():
```
Comment (hannes-ucsc) [inline src/toil/job.py]: This whole while loop should be a method in HarbingerFile.

File: src/toil/job.py [hunk 15]
```diff
@@ -1176,13 +1225,13 @@ def readGlobalFile(self, fileStoreID, userPath=None, cache=True, mutable=None):
                 # FileStoreID exists.  If it does, the wait and periodically check for the removal
                 # of the file and the addition of the completed download into cache of the file by
                 # the other job. Then we link to it.
-                elif fileIsLocal and os.path.exists(harbingerFileName):
+                elif fileIsLocal and harbingerFile.exists():
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Consider

1) turning the `and` into a nested if
2) combining the nested if and the while

Only works if you're ok with producing the log message every 20s.

File: src/toil/job.py [hunk 16]
```diff
@@ -1077,6 +1116,16 @@ def writeGlobalFile(self, localFileName, cleanup=False):
                     fileHandle = open(absLocalFileName, 'r')
                     with self._pendingFileWritesLock:
                         self._pendingFileWrites.add(jobStoreFileID)
+                    # Before we can start the async process, we should also create a dummy harbinger
+                    # file in the cache such that any subsequent jobs asking for this file will not
+                    # attempt to download it from the job store till the write is complete.  We do
+                    # this now and then overwrite the harbinger with the true pid of the writing
+                    # thread because there is an edge case where readGlobalFile in a subsequent job
```
Comment (hannes-ucsc) [inline src/toil/job.py]: This verbiage is out of date/misleading now that we established that all threads share the main threads PID.

**Comment by hannes-ucsc** [comment]
This is already pretty good. My main concern is that we can move more code into HarbingerFile.

Now that we identified a class, we should strive to hide the the habinger file logic from the rest of the program. Encapsulation!

File: src/toil/job.py (hunk 6 above)
Comment (arkal) [inline src/toil/job.py]: That's a really awesome way to represent the XOR! Nice!

File: src/toil/job.py (hunk 15 above)
Comment (arkal) [inline src/toil/job.py]: It messes up the conditionals then. The else in the if harbingerFile.exists is the same as the else in the outer block

**Comment by arkal** [comment]
@hannes-ucsc i implemented the suggested changes

**Comment by hannes-ucsc** [comment]
Well done.

**Comment by arkal** [comment]
Thanks!

**Pull request merged**